src/Microsoft.AspNet.Razor/CodeGenerators/CSharpTagHelperCodeRenderer.cs

```diff
@@ -172,75 +166,37 @@ protected virtual string GenerateUniqueId()
-            var confirmedDictionaries = new HashSet<string>(StringComparer.Ordinal);
-
-            // First attribute wins, even if there are duplicates.
-            var distinctAttributes = chunkAttributes.Distinct(KeyValuePairKeyComparer.Default);
-
-            // Go through the HTML attributes in source order, assigning to properties or indexers as we go.
-            foreach (var attributeKeyValuePair in distinctAttributes)
-            {
-                var attributeName = attributeKeyValuePair.Key;
-                var attributeValueChunk = attributeKeyValuePair.Value;
-                if (attributeValueChunk == null)
+                // Execution contexts and thrown errors are a runtime feature.
+                if (_designTimeMode)
                 {
-                    // Minimized attributes are not valid for bound attributes. TagHelperBlockRewriter has already
-                    // logged an error if it was a bound attribute; so we can skip.
                     continue;
                 }
 
-                // Find the matching TagHelperAttributeDescriptor.
-                var attributeDescriptor = attributeDescriptors.FirstOrDefault(
-                    descriptor => descriptor.IsNameMatch(attributeName));
-                if (attributeDescriptor == null)
-                {
-                    // Attribute is not bound to a property or indexer in this tag helper.
-                    continue;
-                }
+                _writer.WriteInstanceMethodInvocation(
+                    ExecutionContextVariableName,
+                    _tagHelperContext.ExecutionContextAddMethodName,
+                    tagHelperVariableName);
 
-                // We capture the tag helper's property value accessor so we can retrieve it later (if we need to).
-                var valueAccessor = string.Format(
-                    CultureInfo.InvariantCulture,
-                    "{0}.{1}",
-                    tagHelperVariableName,
-                    attributeDescriptor.PropertyName);
+                // Track dictionary properties we have confirmed are non-null.
+                var confirmedDictionaries = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
 
-                if (attributeDescriptor.IsIndexer)
+                // Ensure that all created TagHelpers have initialized dictionary bound properties which are used
+                // via TagHelper indexers.
+                foreach (var chunkAttribute in chunk.Attributes)
                 {
-                    // Need a different valueAccessor in this case. But first need to throw a reasonable Exception at
-                    // runtime if the property is null. The check is not required at design time.
-                    if (!_designTimeMode && confirmedDictionaries.Add(attributeDescriptor.PropertyName))
+                    var associatedAttributeDescriptor = tagHelperDescriptor.Attributes.FirstOrDefault(
```

Comment:
Add a `continue` if `chunkAttribute.Value == null` since no indexer assignment will occur. Also make sure this case is tested. If it's actually impossible to reach here with a bound attribute that has `chunkAttribute.Value == null`, turn the `null` check at the top of `RenderBoundAttribute()` into a `Debug.Assert()` and use same thing here.